### ISSUE TITLE
The notebook's code auto-completion feature is displaying inconsistency in its functionality.

### ISSUE THREAD

**Issue by borah17** [jupyterlab/jupyterlab#15110]
As we are using notebook template code from  jupyterlab example in our project. We've encountered an issue with auto completion of code in the notebook. Sometimes it works and sometimes it does not. Could you provide insights into the potential reasons for this behavior?
Note*: Currently we're using 4.0.4 version of jupyterlab. The same issue persist for previous version also.

**Comment by krassowski** [comment]
It would really help if you could narrow down "sometimes". One think to check is whether a kernel is connected or not (if it is disconnected it cannot provide completions). In general more details would be appreciated (browser, kernel, and anything that the [bug issue template](https://github.com/jupyterlab/jupyterlab/issues/new?assignees=&labels=bug%2C+status%3ANeeds+Triage&projects=&template=bug_report.md) asks for).

**Comment by krassowski** [comment]
Also, do you refer to tab-completion or automatic invocation of the completer, i.e. did you modify any settings (menu → Settings → Completer, can you take a screenshot)?

**Comment by francisconog** [comment]
I'm facing a similar situation. In my case the autocomplete shows after using `:` or after executing a cell with `shift + enter`. This doen't happens in NbClassic and it sloows down my typing (and annoys a lot). I'm using notebook==7.0.3
<img width="959" alt="image" src="https://github.com/jupyterlab/jupyterlab/assets/43391120/47ef4ed8-bf5d-465f-9efb-708b49a9115f">

**Comment by krassowski** [comment]
In absence to replies to the questions above, I assume this is a duplicate of https://github.com/jupyterlab/jupyterlab/issues/15013. Should we remove the default auto-invocation heuristic until it is refined @trungleduc? There was a comment saying:

> I'm facing a similar situation. In my case the autocomplete shows after using : or after executing a cell with shift + enter. This doen't happens in NbClassic and it sloows down my typing (and annoys a lot). I'm using notebook==7.0.3

but it appears to have been removed.

**Comment by trungleduc** [comment]
>  Should we remove the default auto-invocation heuristic until it is refined

Completely agree!

**Comment by krassowski** [comment]
This should probably be addressed together with https://github.com/jupyterlab/jupyterlab/issues/15022 so that downstream can still enable it if they provide a good heuristic.

**Comment by laser-vis** [comment]
> In absence to replies to the questions above, I assume this is a duplicate of #15013. Should we remove the default auto-invocation heuristic until it is refined @trungleduc? There was a comment saying:
> 
> > I'm facing a similar situation. In my case the autocomplete shows after using : or after executing a cell with shift + enter. This doen't happens in NbClassic and it sloows down my typing (and annoys a lot). I'm using notebook==7.0.3
> 
> but it appears to have been removed.

I have the exact issue the author of the removed comment reported. Autocomplete shows up (not always, but more often than not) upon executing cells - using shift+enter, ctrl+enter, or the run button in toolbar. It also shows up "a lot" of other times unexpectedly and grabs keyboard focus, e.g. when typing and hitting return at the end of a line, autocomplete somehow gets invoked and the "return" selects and adds some unexpected "helpful suggestion" to whatever I was typing. I have had to turn-off code-completion to avoid these annoyances.

I have been looking for reports of similar behavior in open issues etc, but didn't find any and assumed there was something wrong with my particular installation. It appears at least one other person has experiences something similar.

I will try and post a gif or at least some screenshots of the buggy behavior.

**Comment by krassowski** [comment]
Thank you for extra details @laser-vis - please do post a GIF and at least a screenshot of your Code Completion settings and any logs there may be in the browser developer console - this would be greatly appreciated. Are you using Notebook or JupyterLab? Which version?

**Comment by trungleduc** [comment]
> This should probably be addressed together with #15022 so that downstream can still enable it if they provide a good heuristic.

Actually, I'm running out of cycles to put on JupyterLab till the end of the year, will try to tackle this issue later.

**Comment by laser-vis** [comment]
Will provide more details this afternoon; apologies for not having done so sooner...
________________________________

**Comment by laser-vis** [comment]
Here are some details from my installation:
- Browser: Edge v116
- JupyterLab: v4.0.5
- Notebook: v7.0.3
- jupyterlab-lsp: v5.0.0b

Code-completion settings:

**Comment by krassowski** [comment]
> jupyterlab-lsp: v5.0.0b

Well, that came with a pretty clear warning that completer is broken:

![image](https://github.com/jupyterlab/jupyterlab/assets/5832902/0e7c0562-30f9-407c-934a-21601a087168)

The relevant things were fixed in a combination of JupyterLab: v4.0.6 + jupyterlab-lsp [v5.0.0rc0](https://github.com/jupyter-lsp/jupyterlab-lsp/releases/tag/v5.0.0rc0). Please upgrade and see how it works with the latest versions.

**Comment by laser-vis** [comment]
I will certainly do that @krassowski , but as I note in my updated comment above, I was seeing the same behavior prior to jupyterlab-lsp installation.

**Comment by krassowski** [comment]
Thank you, these are useful things to test when fixing https://github.com/jupyterlab/jupyterlab/issues/15013. For now using latest version of LSP and JupyterLab does help as it ships with better auto completion heuristics. This will also help in Notebook once a new version is released.

**Comment by borah17** [comment]
> Also, do you refer to tab-completion or automatic invocation of the completer, i.e. did you modify any settings (menu → Settings → Completer, can you take a screenshot)?

Yes, I'm referring to the tab-completion  also there was no change in the settings

**Comment by krassowski** [comment]
@borah17 so your issue appears to be different from @laser-vis' issue. Can you please provide more information on **when** the tab-completion does not work? Can you also check latest version (4.0.6)?

**Comment by JasonWeill** [comment]
@borah17 Any additional information, as requested by @krassowski ?

**Comment by borah17** [comment]
> @borah17 Any additional information, as requested by @krassowski ?

I've tried with 4.0.6 but getting the same result.
![code-completion-issue](https://github.com/jupyterlab/jupyterlab/assets/97163158/f3155d44-1c57-4025-96db-4bd50a466fa0)

**Comment by krassowski** [comment]
When are you pressing tab in the recording?

**Comment by borah17** [comment]
> When are you pressing tab in the recording?

I press the Tab when I input a dot (.) following the os variable, you can see it gives space instead of the code completion popup. Sometimes it works sometimes it does not.
